src/QsCompiler/TextProcessor/CompletionParsing.fs

```diff
@@ -5,49 +5,110 @@
 /// closed if it is still valid for the bracket to be closed later.
 module Microsoft.Quantum.QsCompiler.TextProcessing.CompletionParsing
 
+#nowarn "40"
+
 open System
 open FParsec
+open Microsoft.Quantum.QsCompiler.TextProcessing.ExpressionParsing
 open Microsoft.Quantum.QsCompiler.TextProcessing.Keywords
 open Microsoft.Quantum.QsCompiler.TextProcessing.ParsingPrimitives
 open Microsoft.Quantum.QsCompiler.TextProcessing.SyntaxBuilder
-open Microsoft.Quantum.QsCompiler.TextProcessing.TypeParsing
 
 
 /// Describes the environment of a code fragment in terms of what kind of completions are available.
 type CompletionEnvironment =
     /// The code fragment is inside a namespace but outside any callable.
     | NamespaceTopLevel
+    /// The code fragment is a statement inside a callable.
+    | Statement
 
 /// Describes the kind of identifier that is expected at a particular position in the source code.
 type IdentifierKind =
-    /// The identifier is a new symbol declaration.
-    | Declaration
-    /// The identifier is a type name.
-    | Type
-    /// The identifier is a characteristic set name.
-    | Characteristic
     /// The identifier is the given keyword.
     | Keyword of string
+    /// The identifier is a variable or callable name.
+    | Variable
+    /// The identifier is a user-defined type name.
+    | UserDefinedType
+    /// The identifier is a type parameter.
+    | TypeParameter
+    /// The identifier is a new symbol declaration.
+    | Declaration
     /// The identifier is a namespace.
     | Namespace
     /// The identifier is a member of the given namespace and has the given kind.
     | Member of string * IdentifierKind
+    /// The identifier is a named item in a user-defined type.
+    // TODO: Add information so completion knows the type being accessed.
+    | NamedItem
+
+/// The exception that is thrown when the completion parser can't parse the text of a fragment.
+exception CompletionParserError of detail : string * error : ParserError with
+    override this.Message = this.error.ToString()
+    override this.ToString() = this.detail
+
+/// Merges the error messages from all replies in the list.
+let rec private toErrorMessageList (replies : Reply<'a> list) =
+    match replies with
+    | reply :: tail when reply.Status = Ok -> toErrorMessageList tail
+    | reply :: tail -> ErrorMessageList.Merge (reply.Error, toErrorMessageList tail)
+    | [] -> null
+
+/// The missing expression keyword.
+let private missingExpr =
+    { parse = keyword "_"; id = "_" }
```

Comment:
`missingExpr` in ParsingPrimitives.fs discards the position info, so I would have to wrap it in a term again: `parse = missingExpr |> term |>> snd`. Returning a default position might also work since I don't think it's used anywhere. But since I'm already duplicating the keyword ID, duplicating the keyword parser (which has the right type) didn't seem too bad... which do you think is best?

Reply:
I didn't realize - sorry, didn't check carefully enough. I'm fine with the duplication.